Model and ViewModel for message-list feature.

BarAgam opened the pull request:
Made the ViewModel and the Model parts of the messages management use-case (for messages displaying, choosing selected message, and read status of different messages).
MessagesViewModel needs MessagesModel. MessagesReadStatusModel and SelectedMessageModel to function.
The models have been implemented with an in-memory implementation. The app will use these implementations, until a better alternative (DataBase for exmaple) will be implemented. The tests use the in-memory implementations, even if another implementation will be created, since they're perfect for testing.

Both the InMemoryMessageModel and MessagesViewModel class have test-classes. The other classes are mostly data objects/ simple enough not to be tested.

barmako committed:
Made the ViewModel and the Model parts of the messages management use-case (for messages displaying, choosing selected message, and read status of different messages).
MessagesViewModel needs MessagesModel. MessagesReadStatusModel and SelectedMessageModel to function.
The models have been implemented with an in-memory implementation. The app will use these implementations, until a better alternative (DataBase for exmaple) will be implemented. The tests use the in-memory implementations, even if another implementation will be created, since they're perfect for testing.

Both the InMemoryMessageModel and MessagesViewModel class have test-classes. The other classes are mostly data objects/ simple enough not to be tested.

review-ninja commented:
<a href="https://app.review.ninja/teamagam/GimelGimel/pull/56" target="_blank"><img src="https://app.review.ninja/52010471/pull/56/badge" alt="ReviewNinja"/>

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/common/DataChangedObservable.java:
```diff
@@ -0,0 +1,8 @@
+package com.teamagam.gimelgimel.app.common;
```
dont you think that this should be in model dir?

BarAgam replied:
Its a common class.

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/common/DataChangedObserver.java:
```diff
@@ -0,0 +1,8 @@
+package com.teamagam.gimelgimel.app.common;
```
same

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/common/NotifyingDataChangedObservable.java:
```diff
@@ -0,0 +1,21 @@
+package com.teamagam.gimelgimel.app.common;
```
...

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/common/NotifyingDataChangedObservable.java:
```diff
@@ -0,0 +1,21 @@
+    private DataChangedObserver mObserver;
+
+    @Override
+    public void setObserver(DataChangedObserver observer) {
```
where do you manage list of observers?

BarAgam replied:
!fix

BarAgam replied:
Sababa Achi

BarAgam replied:
!fixed

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/Message.java:
```diff
@@ -43,40 +43,19 @@ public T getContent() {
         return mContent;
     }
 
-    public void setContent(T content) {
```
why is that?? we still need some of these functionality.

BarAgam replied:
None of them are used, and that's the way it should be. We wouldn't want anyone to change the type/content, id, sender id or creation date  of an existing message.

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/DisplayMessage.java:
```diff
@@ -0,0 +1,47 @@
+public class DisplayMessage {
+
+    private Message mMessage;
+    private boolean mIsSelected;
```
HaHa mIsRead for each message ;)

teamagam1 replied:
I mean mIsSelected

teamagam1 replied:
OK i got it, never mind

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/Message.java:
```diff
@@ -43,40 +43,19 @@ public T getContent() {
         return mCreatedAt;
     }
 
-    public void setCreatedAt(Date createdAt) {
```
for example

BarAgam replied:
Its generated by the server.

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModel.java:
```diff
@@ -0,0 +1,95 @@
+/**
+ * Messages view-model for messages presentation use-case
+ */
+public class MessagesViewModel extends NotifyingDataChangedObservable implements DataChangedObservable {
```
I dont like that this VM extends NotifyingDataChangedObservable.

consider changing It's name to BaseViewModel !resolve

BarAgam replied:
But its not a base viewmodel class.
" I dont like" is not a valid argument. !resolved

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModel.java:
```diff
@@ -0,0 +1,95 @@
+                }
+
+                mSelectedMessageModel.select(message);
+                return index == 0;
```
IMO should be return true. !fix

BarAgam replied:
It shouldn't. It should return true for the first message (or some other default message if you can think of anything more reasonable) but not all messages should be marked as read. !fixed

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/model/entities/messages/InMemory/InMemoryMessagesModel.java:
```diff
@@ -0,0 +1,43 @@
+ */
+public class InMemoryMessagesModel extends NotifyingDataChangedObservable implements MessagesModel {
+
+    private ArrayList<Message> mMessages = new ArrayList<>();
```
those model classes should be singeltons

teamagam1 replied:
!resolve

BarAgam replied:
They dont need to be singletons. We'll only create one instance of them. !resolved

teamagam1 commented on app/src/test/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModelTest.java:
```diff
@@ -0,0 +1,185 @@
+        MessagesViewModel.DisplayedMessagesRandomAccessor res = mMessageViewModel.getDisplayedMessagesRandomAccessor();
+
+        //Assert
+        assertThat(res.get(0).isSelected(), is(true));
```
what should happen if I do the followin test:
        assertThat(res.get(1).isSelected(), is(true));
        assertThat(res.get(0).isSelected(), is(false));
I think this is OK

BarAgam replied:
fixed that.

teamagam1 commented on app/src/test/java/com/teamagam/gimelgimel/app/model/entities/messages/InMemoryMessagesModelTest.java:
```diff
@@ -0,0 +1,144 @@
+        mMessageModel.remove(idx2);
+
+        //Assert
+        assertThat(mMessageModel.get(idx1), sameInstance(mMessage));
```
the indices shouldnt be consistent at all!

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModel.java:
```diff
@@ -0,0 +1,95 @@
+    private SelectedMessageModel mSelectedMessageModel;
+    private MessagesReadStatusModel mMessagesReadStatusModel;
+
+    public MessagesViewModel(
```
How the VM notifies the View when data is updated? !resolve

teamagam1 replied:
How the VM notifies the View when data is updated? !resolve

BarAgam replied:
!resolved

BarAgam replied:
Through the NotifyingDataChangedObservable

BarAgam commented on app/src/main/java/com/teamagam/gimelgimel/app/common/NotifyingDataChangedObservable.java:
```diff
@@ -0,0 +1,21 @@
+        mObserver = observer;
+    }
+
+    public void notifyObserver(){
```
change to protected !fix

BarAgam replied:
!fixed

barmako committed:
Changed observable to support multiple observers instead of one.
Fixed a bug in MessagesModelView getAccessor... implementation where for a message model that doesn't contain a selected message it selects all messages.
Added a test to prevent this bug from reoccurring.

barmako committed:
Intergation of Model-ModelView-View is completed !

barmako committed:
MessagesViewModel will update model's selected and message read status only when it actually changes something. (meaning, a message already marked as read/select will cause VM to update model)

barmako committed:
DisplayMessage constructor  made private, only builder should be used.
Removed unused methods
